File: swift-app/Sources/HuntingMapGenerator/Storage.swift
import Foundation

// MARK: - AppSettings

/// Mirrors the Rust AppSettings struct (serde camelCase).
/// Property names are already camelCase so JSONEncoder uses them as-is.
struct AppSettings: Codable {
    var version: Int
    var theme: String
    var units: String
    var difficulty: String
    var dataLocation: String?
    var onlineByDefault: Bool
    var snapshotRetention: Int
    var currencyWarningMonths: Int
    var defaultDpi: Int
    var autoOpenFolderAfterExport: Bool

    static let defaults = AppSettings(
        version: 1,
        theme: "system",
        units: "imperial",
        difficulty: "beginner",
        dataLocation: nil,
        onlineByDefault: true,
        snapshotRetention: 20,
        currencyWarningMonths: 12,
        defaultDpi: 200,
        autoOpenFolderAfterExport: true
    )

    /// Mirrors `#[serde(default)]` on the Rust struct. The synthesised decoder
    /// would reject a settings.json missing any key, and `baseDataDir` treats a
    /// decode failure as "no custom location" — so a hand-edited file with one
    /// key dropped would make a relocated data directory, and every project in
    /// it, appear to vanish. Decoding each key independently avoids that and
    /// keeps older settings files loading as fields are added.
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let d = AppSettings.defaults
        version                   = try c.decodeIfPresent(Int.self,    forKey: .version)                   ?? d.version
        theme                     = try c.decodeIfPresent(String.self, forKey: .theme)                     ?? d.theme
        units                     = try c.decodeIfPresent(String.self, forKey: .units)                     ?? d.units
        difficulty                = try c.decodeIfPresent(String.self, forKey: .difficulty)                ?? d.difficulty
        dataLocation              = try c.decodeIfPresent(String.self, forKey: .dataLocation)
        onlineByDefault           = try c.decodeIfPresent(Bool.self,   forKey: .onlineByDefault)           ?? d.onlineByDefault
        snapshotRetention         = try c.decodeIfPresent(Int.self,    forKey: .snapshotRetention)         ?? d.snapshotRetention
        currencyWarningMonths     = try c.decodeIfPresent(Int.self,    forKey: .currencyWarningMonths)     ?? d.currencyWarningMonths
        defaultDpi                = try c.decodeIfPresent(Int.self,    forKey: .defaultDpi)                ?? d.defaultDpi
        autoOpenFolderAfterExport = try c.decodeIfPresent(Bool.self,   forKey: .autoOpenFolderAfterExport) ?? d.autoOpenFolderAfterExport
    }

    /// Explicit memberwise init, since providing `init(from:)` suppresses the
    /// synthesised one that `defaults` above relies on.
    init(version: Int, theme: String, units: String, difficulty: String,
         dataLocation: String?, onlineByDefault: Bool, snapshotRetention: Int,
         currencyWarningMonths: Int, defaultDpi: Int, autoOpenFolderAfterExport: Bool) {
        self.version = version
        self.theme = theme
        self.units = units
        self.difficulty = difficulty
        self.dataLocation = dataLocation
        self.onlineByDefault = onlineByDefault
        self.snapshotRetention = snapshotRetention
        self.currencyWarningMonths = currencyWarningMonths
        self.defaultDpi = defaultDpi
        self.autoOpenFolderAfterExport = autoOpenFolderAfterExport
    }
}

// MARK: - Storage

enum Storage {

    // MARK: - Directories

    /// ~/Library/Application Support/com.huntingmapgenerator.app
    static var defaultAppDir: URL {
        let fm = FileManager.default
        let appSupport = fm.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        return appSupport.appendingPathComponent("com.huntingmapgenerator.app")
    }

    static var settingsFileURL: URL {
        defaultAppDir.appendingPathComponent("settings/settings.json")
    }

    /// Returns the active base data directory.
    /// Reads `dataLocation` from settings.json; falls back to defaultAppDir
    /// when the setting is absent or the path no longer exists.
    static func baseDataDir() throws -> URL {
        let fm = FileManager.default
        if fm.fileExists(atPath: settingsFileURL.path),
           let data = try? Data(contentsOf: settingsFileURL),
           let settings = try? JSONDecoder().decode(AppSettings.self, from: data),
           let custom = settings.dataLocation {
            let customURL = URL(fileURLWithPath: custom)
            if fm.fileExists(atPath: customURL.path) {
                return customURL
            }
        }
        return defaultAppDir
    }

    /// Creates the seven canonical subdirectories and a default settings.json
    /// on first launch.  Safe to call on every launch.
    static func initDirectories() throws {
        let fm = FileManager.default
        let base = defaultAppDir

        for sub in ["projects", "data", "settings", "presets", "regions", "access-rules", "logs"] {
            try fm.createDirectory(
                at: base.appendingPathComponent(sub),
                withIntermediateDirectories: true
            )
        }

        if !fm.fileExists(atPath: settingsFileURL.path) {
            try writeSettings(AppSettings.defaults)
        }
    }

    // MARK: - IPC commands

    static func getDataDir() throws -> String {
        return try baseDataDir().path
    }

    /// Returns settings as a JSON-compatible dictionary (for IPCHandler).
    static func getSettings() throws -> [String: Any] {
        let fm = FileManager.default
        let data: Data
        if fm.fileExists(atPath: settingsFileURL.path) {
            data = try Data(contentsOf: settingsFileURL)
        } else {
            data = try encodeSettings(AppSettings.defaults)
        }
        guard let dict = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw StorageError.jsonParseFailed("settings.json is not a JSON object")
        }
        return dict
    }

    /// Merges one key/value pair into settings.json.
    static func setSetting(key: String, value: Any) throws {
        let fm = FileManager.default
        var dict: [String: Any]

        if fm.fileExists(atPath: settingsFileURL.path) {
            let data = try Data(contentsOf: settingsFileURL)
            guard let parsed = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                throw StorageError.jsonParseFailed("settings.json is not a JSON object")
            }
            dict = parsed
        } else {
            let data = try encodeSettings(AppSettings.defaults)
            dict = try JSONSerialization.jsonObject(with: data) as! [String: Any]
        }

        dict[key] = value
        let updated = try JSONSerialization.data(withJSONObject: dict, options: .prettyPrinted)
        try updated.write(to: settingsFileURL)
    }

    // MARK: - Private helpers

    private static func encodeSettings(_ s: AppSettings) throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = .prettyPrinted
        return try encoder.encode(s)
    }

    private static func writeSettings(_ s: AppSettings) throws {
        let data = try encodeSettings(s)
        try data.write(to: settingsFileURL)
    }

    // MARK: - App log (Stage 22)

    static var logFileURL: URL {
        defaultAppDir.appendingPathComponent("logs/app.log")
    }

    static func writeAppLog(message: String) throws {
        let fm = FileManager.default
        let dir = logFileURL.deletingLastPathComponent()
        try fm.createDirectory(at: dir, withIntermediateDirectories: true)

        let fmt = ISO8601DateFormatter()
        fmt.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let ts = fmt.string(from: Date())
        let line = "[\(ts)] \(message)\n"

        var content = ""
        if fm.fileExists(atPath: logFileURL.path) {
            content = (try? String(contentsOf: logFileURL, encoding: .utf8)) ?? ""
        }
        content += line

        // Trim to 2000 lines
        let lines = content.components(separatedBy: "\n")
        let trimmed = lines.count > 2000
            ? lines.suffix(2000).joined(separator: "\n") + "\n"
            : content
        try trimmed.write(to: logFileURL, atomically: true, encoding: .utf8)
    }

    static func readAppLog(lines: Int) throws -> String {
        guard FileManager.default.fileExists(atPath: logFileURL.path) else { return "" }
        let content = try String(contentsOf: logFileURL, encoding: .utf8)
        let all = content.components(separatedBy: "\n")
        let slice = all.count > lines ? Array(all.suffix(lines)) : all
        return slice.joined(separator: "\n")
    }
}

// MARK: - Errors

enum StorageError: LocalizedError {
    case jsonParseFailed(String)

    var errorDescription: String? {
        switch self {
        case .jsonParseFailed(let msg): return "Storage JSON parse error: \(msg)"
        }
    }
}

File: src-tauri/src/storage.rs
use std::fs;
use std::path::PathBuf;

use chrono::Utc;
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Manager};

// ---------------------------------------------------------------------------
// Settings schema
// ---------------------------------------------------------------------------

/// `serde(default)` matters here beyond tidiness: settings.json is documented as
/// hand-editable, and `base_data_dir` silently falls back to the default
/// location when this fails to parse. Without per-field defaults, one missing
/// key would make a relocated data directory — and every project in it — appear
/// to vanish. It also keeps older settings files loading as fields are added.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct AppSettings {
    pub version: u32,
    pub theme: String,                   // "light" | "dark" | "system"
    pub units: String,                   // "imperial" | "metric"
    pub difficulty: String,              // "beginner" | "intermediate" | "advanced"
    pub data_location: Option<String>,   // null → use OS app-data dir
    pub online_by_default: bool,
    pub snapshot_retention: u32,         // number of snapshots to keep
    pub currency_warning_months: u32,    // flag data older than this
    pub default_dpi: u32,               // 150 | 200 | 300 | 400
    pub auto_open_folder_after_export: bool,
}

impl Default for AppSettings {
    fn default() -> Self {
        Self {
            version: 1,
            theme: "system".into(),
            units: "imperial".into(),
            difficulty: "beginner".into(),
            data_location: None,
            online_by_default: true,
            snapshot_retention: 20,
            currency_warning_months: 12,
            default_dpi: 200,
            auto_open_folder_after_export: true,
        }
    }
}

// ---------------------------------------------------------------------------
// Path helpers
// ---------------------------------------------------------------------------

/// Returns the base data directory.
/// Always reads from the settings.json in the DEFAULT app-data dir first,
/// so we know where to look even when the user has relocated data.
pub fn base_data_dir(app: &AppHandle) -> Result<PathBuf, String> {
    let default_dir = app
        .path()
        .app_data_dir()
        .map_err(|e| format!("Cannot determine app data directory: {e}"))?;

    let settings_path = default_dir.join("settings").join("settings.json");
    if settings_path.exists() {
        if let Ok(text) = fs::read_to_string(&settings_path) {
            if let Ok(s) = serde_json::from_str::<AppSettings>(&text) {
                if let Some(custom) = s.data_location {
                    let p = PathBuf::from(&custom);
                    if p.exists() {
                        return Ok(p);
                    }
                }
            }
        }
    }

    Ok(default_dir)
}

// ---------------------------------------------------------------------------
// Startup initialisation
// ---------------------------------------------------------------------------

/// Creates the full directory tree and a default settings.json on first run.
/// Safe to call on every launch — already-existing dirs are skipped.
pub fn init_directories(app: &AppHandle) -> Result<(), Box<dyn std::error::Error>> {
    let base = app.path().app_data_dir()?;

    for sub in &[
        "projects",
        "data",
        "settings",
        "presets",
        "regions",
        "access-rules",
        "logs",
    ] {
        fs::create_dir_all(base.join(sub))?;
    }

    // Write default settings.json only on first launch
    let settings_path = base.join("settings").join("settings.json");
    if !settings_path.exists() {
        let json = serde_json::to_string_pretty(&AppSettings::default())?;
        fs::write(&settings_path, json)?;
    }

    Ok(())
}

// ---------------------------------------------------------------------------
// Tauri commands
// ---------------------------------------------------------------------------

#[tauri::command]
pub fn get_data_dir(app: AppHandle) -> Result<String, String> {
    base_data_dir(&app).map(|p| p.to_string_lossy().into_owned())
}

#[tauri::command]
pub fn get_settings(app: AppHandle) -> Result<AppSettings, String> {
    let base = app
        .path()
        .app_data_dir()
        .map_err(|e| e.to_string())?;
    let path = base.join("settings").join("settings.json");

    if path.exists() {
        let text = fs::read_to_string(&path).map_err(|e| e.to_string())?;
        serde_json::from_str(&text).map_err(|e| e.to_string())
    } else {
        Ok(AppSettings::default())
    }
}

// ---------------------------------------------------------------------------
// App log (Stage 22)
// ---------------------------------------------------------------------------

/// Appends a timestamped line to logs/app.log (max 2000 lines retained).
#[tauri::command]
pub fn write_app_log(app: AppHandle, message: String) -> Result<(), String> {
    let base = app.path().app_data_dir().map_err(|e| e.to_string())?;
    let log_path = base.join("logs").join("app.log");
    fs::create_dir_all(log_path.parent().unwrap()).map_err(|e| e.to_string())?;

    let ts = Utc::now().to_rfc3339();
    let line = format!("[{ts}] {message}\n");

    use std::io::{Read, Write};
    let mut content = String::new();
    if log_path.exists() {
        let mut f = std::fs::File::open(&log_path).map_err(|e| e.to_string())?;
        f.read_to_string(&mut content).map_err(|e| e.to_string())?;
    }
    content.push_str(&line);

    // Trim to last 2000 lines
    let lines: Vec<&str> = content.lines().collect();
    let trimmed = if lines.len() > 2000 {
        lines[lines.len() - 2000..].join("\n") + "\n"
    } else {
        content
    };

    let mut f = std::fs::File::create(&log_path).map_err(|e| e.to_string())?;
    f.write_all(trimmed.as_bytes()).map_err(|e| e.to_string())
}

/// Returns the last N lines of the app log (default 200).
#[tauri::command]
pub fn read_app_log(app: AppHandle, lines: Option<usize>) -> Result<String, String> {
    let base = app.path().app_data_dir().map_err(|e| e.to_string())?;
    let log_path = base.join("logs").join("app.log");
    if !log_path.exists() {
        return Ok(String::new());
    }
    let content = fs::read_to_string(&log_path).map_err(|e| e.to_string())?;
    let n = lines.unwrap_or(200);
    let all: Vec<&str> = content.lines().collect();
    let slice = if all.len() > n { &all[all.len() - n..] } else { &all[..] };
    Ok(slice.join("\n"))
}

/// Merges a single key/value pair into settings.json.
#[tauri::command]
pub fn set_setting(
    app: AppHandle,
    key: String,
    value: serde_json::Value,
) -> Result<(), String> {
    let base = app
        .path()
        .app_data_dir()
        .map_err(|e| e.to_string())?;
    let path = base.join("settings").join("settings.json");

    let mut doc: serde_json::Value = if path.exists() {
        let text = fs::read_to_string(&path).map_err(|e| e.to_string())?;
        serde_json::from_str(&text).map_err(|e| e.to_string())?
    } else {
        serde_json::to_value(AppSettings::default()).map_err(|e| e.to_string())?
    };

    if let serde_json::Value::Object(ref mut map) = doc {
        map.insert(key, value);
    }

    let json = serde_json::to_string_pretty(&doc).map_err(|e| e.to_string())?;
    fs::write(&path, json).map_err(|e| e.to_string())?;

    Ok(())
}

#[cfg(test)]
mod settings_tests {
    use super::*;

    /// settings.json is documented as hand-editable, and a parse failure makes
    /// `base_data_dir` silently ignore a relocated data directory. A file
    /// missing keys must therefore still load.
    #[test]
    fn partial_settings_file_loads_with_defaults() {
        let json = r#"{"dataLocation":"/Volumes/Maps","units":"metric"}"#;
        let s: AppSettings = serde_json::from_str(json).unwrap();
        assert_eq!(s.data_location.as_deref(), Some("/Volumes/Maps"),
            "a relocated data directory must survive a partial file");
        assert_eq!(s.units, "metric");
        // everything absent falls back to the defaults
        assert_eq!(s.theme, "system");
        assert_eq!(s.snapshot_retention, 20);
        assert_eq!(s.default_dpi, 200);
    }

    /// An empty object is still valid and yields defaults throughout.
    #[test]
    fn empty_settings_file_loads_defaults() {
        let s: AppSettings = serde_json::from_str("{}").unwrap();
        assert_eq!(s.theme, AppSettings::default().theme);
        assert_eq!(s.data_location, None);
    }

    /// Unknown keys (set_setting writes app-specific ones like
    /// hide_access_disclaimer) must not break parsing.
    #[test]
    fn unknown_keys_are_ignored() {
        let json = r#"{"theme":"dark","hide_access_disclaimer":true,"somethingNew":42}"#;
        let s: AppSettings = serde_json::from_str(json).unwrap();
        assert_eq!(s.theme, "dark");
    }

    /// A full round-trip still works, so writing then reading is stable.
    #[test]
    fn round_trips_full_settings() {
        let original = AppSettings {
            data_location: Some("/tmp/x".into()),
            snapshot_retention: 5,
            ..AppSettings::default()
        };
        let text = serde_json::to_string(&original).unwrap();
        let back: AppSettings = serde_json::from_str(&text).unwrap();
        assert_eq!(back.data_location, original.data_location);
        assert_eq!(back.snapshot_retention, 5);
    }
}
